File: ketch-android/Android Ketch Pref Center using Java/app/src/main/java/com/ketch/sample/webview/MainActivity.java
package com.ketch.sample.webview;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.google.android.gms.ads.identifier.AdvertisingIdClient;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = MainActivity.class.getSimpleName();

    private static final String ORG_CODE = "<organization code>";
    private static final String PROPERTY = "<property>";
    private static final String ADVERTISING_ID_CODE = "<advertising field code>";

    private KetchSharedPreferences sharedPreferences;

    private KetchWebView ketchWebView;
    private View mainLayout;
    private Button button;
    private ProgressBar progressBar;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        getSupportActionBar().hide();

        sharedPreferences = new KetchSharedPreferences(getApplicationContext());

        ketchWebView = findViewById(R.id.ketchWebView);
        mainLayout = findViewById(R.id.mainLayout);
        button = findViewById(R.id.button);
        progressBar = findViewById(R.id.progressBar);

        ketchWebView.setListener(new KetchWebView.KetchListener() {
            @Override
            public void onCCPAUpdate(String ccpaString) {
                sharedPreferences.saveUSPrivacyString(ccpaString);
            }

            @Override
            public void onTCFUpdate(String tcfString) {
                sharedPreferences.saveTCFTCString(tcfString);
            }

            @Override
            public void onClose() {
                mainLayout.setVisibility(View.VISIBLE);
            }
        });

        button.setOnClickListener(v -> {
            ketchWebView.show();
            mainLayout.setVisibility(View.GONE);
        });

        loadAdvertisingId(advertisingId -> {
            if (advertisingId != null) {
                button.setVisibility(View.VISIBLE);

                // identities to be passed to the WebView displaying the Ketch Preference Center
                List<Identity> identities = new ArrayList<>();
                identities.add(new Identity(ADVERTISING_ID_CODE, advertisingId));

                ketchWebView.init(ORG_CODE, PROPERTY, identities);
            } else {
                button.setVisibility(View.GONE);
            }
        });
    }

    private interface AdvertisingIdCallback {
        void onAdvertisingIdResult(@Nullable String advertisingId);
    }

    private void loadAdvertisingId(AdvertisingIdCallback callback) {
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Handler handler = new Handler(Looper.getMainLooper());

        progressBar.setVisibility(View.VISIBLE);

        executor.execute(() -> {
            String id = null;
            try {
                id = AdvertisingIdClient.getAdvertisingIdInfo(MainActivity.this).getId();
            } catch (Exception e) {
                Log.e(TAG, e.getMessage(), e);
                handler.post(() -> {
                    Toast.makeText(MainActivity.this,
                            R.string.cannot_get_advertising_id_toast,
                            Toast.LENGTH_LONG
                    ).show();
                });
            } finally {
                final String advertisingId = id;
                handler.post(() -> {
                    callback.onAdvertisingIdResult(advertisingId);
                    progressBar.setVisibility(View.GONE);
                });
            }
        });
    }
}

File: ketch-android/webview/Android Ketch Pref Center using Kotlin/app/src/main/java/com/ketch/sample/pref/MainActivity.kt
package com.ketch.sample.pref

import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.isVisible
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.whenStateAtLeast
import com.google.android.gms.ads.identifier.AdvertisingIdClient
import kotlinx.android.synthetic.main.activity_main.button
import kotlinx.android.synthetic.main.activity_main.ketchWebView
import kotlinx.android.synthetic.main.activity_main.mainLayout
import kotlinx.android.synthetic.main.activity_main.progressBar
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch


class MainActivity : AppCompatActivity() {
    private val sharedPreferences: KetchSharedPreferences by lazy {
        KetchSharedPreferences(this)
    }

    private val advertisingId = MutableStateFlow<String?>(null)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        supportActionBar?.hide()

        ketchWebView.listener = object : KetchWebView.KetchListener {
            override fun onCCPAUpdate(ccpaString: String?) {
                sharedPreferences.saveUSPrivacyString(ccpaString)
            }

            override fun onTCFUpdate(tcfString: String?) {
                sharedPreferences.saveTCFTCString(tcfString)
            }

            override fun onClose() {
                mainLayout.isVisible = true
            }
        }

        button.setOnClickListener {
            ketchWebView.show()
            mainLayout.isVisible = false
        }

        collectState(advertisingId) {
            button.isVisible = it != null

            it?.let { aaid ->
                progressBar.isVisible = false

                // identities to be passed to the WebView displaying the Ketch Preference Center
                val identities = ArrayList<Identity>()
                identities.add(Identity(ADVERTISING_ID_CODE, aaid))

                ketchWebView.init(ORG_CODE, PROPERTY, identities)
            }
        }

        progressBar.isVisible = true

        loadAdvertisingId()
    }

    @OptIn(DelicateCoroutinesApi::class)
    private fun loadAdvertisingId() {
        GlobalScope.launch(Dispatchers.IO) {
            try {
                advertisingId.value = AdvertisingIdClient.getAdvertisingIdInfo(applicationContext).id
            } catch (e: Exception) {
                e.printStackTrace()
                launch(Dispatchers.Main) {
                    Toast.makeText(
                        this@MainActivity,
                        R.string.cannot_get_advertising_id_toast,
                        Toast.LENGTH_LONG
                    ).show()
                }
            } finally {
                launch(Dispatchers.Main) {
                    progressBar.isVisible = false
                }
            }
        }
    }

    private fun <A> collectState(
        state: StateFlow<A>,
        minState: Lifecycle.State = Lifecycle.State.STARTED,
        collector: suspend (A) -> Unit
    ) = state.collectLifecycle(minState, collector)

    private fun <T> Flow<T>.collectLifecycle(
        minState: Lifecycle.State,
        collector: suspend (T) -> Unit
    ) {
        lifecycleScope.launch {
            lifecycle.whenStateAtLeast(minState) {
                collect {
                    collector(it)
                }
            }
        }
    }

    companion object {
        private const val ORG_CODE = "ketch_samples"
        private const val PROPERTY = "android"
        private const val ADVERTISING_ID_CODE = "<advertising field code>"
    }
}
